Add ContainerRegistry managed resource

sergenyalcin opened the pull request:
Signed-off-by: sergenyalcin <yalcinsergen97@gmail.com>

### Description of your changes

Fixes #297 

Implementing `ContainerRegistry` managed resource. When a `ContainerRegistry` managed resource is created, it will be ensured that a `Bucket` is created in the specified `location`.

I would like to mention a few issues I encountered while working on this task.

- **Scope of this task**: It is very important to determine the scope correctly in order to see the limits of the implementation and for what purpose this resource will be used. On the Terraform side, a `Bucket` is created when a `ContainerRegistry` resource is created (`Bucket` is required to properly run image actions (_push/pull_) in the `Registry`.) When you delete the `ContainerRegistry` resource you created, the `Bucket` is not deleted (probably to prevent data loss in case this `Bucket` was used for other purposes). Please see https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/container_registry
Currently, this PR implements a similar behavior to Terraform. However, it is possible to extend the scope. Therefore, this is an open discussion.

- **Adding a new dependency**: When you review the Google Cloud Go SDK, there is no rest api implementation related to `ContainerRegistry`. However, I found this repo when I looked through other google's repos-> https://github.com/google/go-containerregistry In order to trigger the `Bucket` creation I mentioned above, it is sufficient to perform a `handshake` with the `gcr.io service`, using the specified project and repo (No need for image push). This handshake operation (without configuring http) is provided via structs and a function in the `go-containerregistry` repo. Therefore, you will see that a new dependency has been added with this PR. Adding this dependency or writing an http request directly are two different ways, which is another point of discussion.

- **Should a new provider be implemented?**: I just mentioned the `go-containerregistry` repo. In this repo, there are many other resources (eg `repository`, `image` etc) api along with `registry`. There is also another `cli`. Therefore, thinking the gcr service as a separate provider may be another option. (cc: @ulucinar )

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

[contribution process]: https://git.io/fj2m9

sergenyalcin commented:
I would like to summarize my own comments on some of the points above. 

First of all, some utility functions of the go-containerregistry repo are used to implement the handshake action in PR. As I mentioned above, this repo has a separate cli and many different api implementations. So this may require it to be implemented as a separate provider. 

Therefore, if we are going to focus on implementing this resource only with this scope on the provider-gcp side (like Terraform) for now, it may be possible to remove this usage and dependency (go-containerregistry) and edit it as an http request.

ulucinar commented:
As we have discussed, I think it makes sense not to consume `go-containerregistry` in `provider-gcp` because:
- As you mentioned, there are many other resources that we can support via `go-containerregistry` such as image tag, etc., which, in our common understanding, is a candidate for a separate provider. Could you please also check if Terraform has a separate provider built on top of `go-containerregistry`?
- Terraform GCP provider does something like this, i.e., makes directly use of an HTTP client to make the `gcr.io` service request to do the initial handshake.

However, if setting up the credentials needed for the HTTP client turns out to be expensive (complex) implementation-wise, than I would prefer to continue with the additional dependency.

ulucinar commented:
Regarding the above comment, we can check whether we can authenticate to GCR using GCloud credentials. That would imply that using the credentials referred by the `ProviderConfig`, we can authenticate against GCR (and maybe easily configure our HTTPClient).

sergenyalcin commented on pkg/controller/registry/containerregistry.go:
```diff
@@ -0,0 +1,202 @@
+	errUpdate           = "cannot update Google Container Registry (GCR) object"
+
+	gcrUrl                          = "gcr.io"
+	testRepoName                    = "crossplane-test"
```
When we try to handshake with `"gcr.io/crossplane-playgroud"` the following error occurs:

> message: 'create failed: an error occurred during handshake: GET https://eu.gcr.io/v2/token?scope=repository%3Acrossplane-playground%3Apush%2Cpull&service=eu.gcr.io: 
UNKNOWN: Pushing to root-level images is disabled. See https://cloud.google.com/container-registry/docs/support/deprecation-notices'

So we need to add a repository name to the end of URL: "gcr.io/crossplane-playgroud/crossplane-test"

Also, please see the Terraform side (`my-repo`): https://github.com/hashicorp/terraform-provider-google/blob/9795b29a7b2157fb82287bafa4880063ebbefcba/google/resource_container_registry.go#L67

turkenh commented on pkg/controller/registry/containerregistry.go:
```diff
@@ -0,0 +1,201 @@
+
+func populateCRFromBucket(cr *v1alpha1.ContainerRegistry, bucket *storage.Bucket) {
+	cr.Status.AtProvider.ID = bucket.Id
+	cr.Status.AtProvider.BucketLink = bucket.SelfLink
```
nit: I would suggest doing this inline above instead of defining a function since it is just two lines.

turkenh commented on pkg/controller/registry/containerregistry.go:
```diff
@@ -0,0 +1,201 @@
+	return bucketName
+}
+
+func (e *external) getBucket(bucketName string) (*storage.Bucket, error) {
```
nit: similarly we can combine `getBucketName` and `getBucket` methods into one, e.g. do what `getBucketName` does as inline in `getBucket`.

turkenh reviewed (commented):
@sergenyalcin thanks a lot for your first contribution to provider-gcp!

Looking very good, just added a couple of comments mostly related to styling and some nitpicks.

sergenyalcin commented on apis/registry/v1alpha1/containerregistry_types.go:
```diff
@@ -0,0 +1,84 @@
+// ContainerRegistryObservation is used to show the observed state of the ContainerRegistry
+type ContainerRegistryObservation struct {
+	// The name of the bucket.
+	ID string `json:"id,omitempty"`
```
I changed the name to BucketID

sergenyalcin commented on pkg/controller/registry/containerregistry.go:
```diff
@@ -0,0 +1,201 @@
+	} else {
+		bucketName = fmt.Sprintf(bucketNameFormatWithoutLocation, e.projectID)
+	}
+	return bucketName
```
The `getBucketName` function was inlined in `getBucket`

sergenyalcin commented:
Hi @turkenh thank you for your review. I addressed your comments.

While testing my PR, I observed that there is an issue about the deletion. Although it is a very basic scenario, I missed it before. The problem is that, the CR could not be deleted because we skipped the deletion and the `resourceExists` field was set to `true` during observation. So I had to add the `deletionTimeStamp` check inside the `Observe`. I also tried to add an explanatory comment:
https://github.com/sergenyalcin/provider-gcp/blob/07048f925b5c25b8f03786aa161617d90cd5532a/pkg/controller/registry/containerregistry.go#L121

Similar reference from provider-azure: https://github.com/crossplane/provider-azure/blob/7e7f436755483ab134b58aeea718e3905501dd90/pkg/controller/database/postgresqlserverconfiguration/managed.go#L128

turkenh reviewed (approved):
Thank you @sergenyalcin, looking great!
